### PR TITLE
[WIP] Verify enough resources for breeze

### LINKED ISSUE #20251: Breeze: Verify if there are enough resources avaiilable in Breeze
At entry of the Breeze command we verify if there iss enouggh CPU/memory/disk space and print (coloured) information if the resources are not enough.

We should replicate that in Python

### PR DESCRIPTION
**Pull request by edithturn**
Check resources before installing docker:
* Memory available
* CPUs available
* Disk available

Using **psutil** to calculate available resources in kilobytes, get_size function will convert in Gb.

I created a dictionary to add resources available and resources needed.

TODO:
* Verbose, colors on the console
* Check if script _in_container_script_init.sh
* Check if for disk this is taking the mounted on root pat "/"
* Check if _docker_engine_resources.sh is calling successful to run_resource_check.py

@potiuk I will appreciate your feedback.

Close: #20251

### CODE REVIEW

File: dev/breeze/doc/BREEZE.md
Comment (potiuk) [inline dev/breeze/doc/BREEZE.md]: Copy * Paste mistake :) ?

File: dev/breeze/doc/BREEZE.md
Comment (potiuk) [inline dev/breeze/doc/BREEZE.md]: Some merge artifacts

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]: I  think we do not really need to source it - it was mainly needed for the bash script to source some common functions but we can simply drop it.

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]: We need to add colors/rich to that :)

**Review by potiuk — CHANGES_REQUESTED** [review]
Few small changes.

File: dev/breeze/doc/BREEZE.md
Comment (edithturn) [reply dev/breeze/doc/BREEZE.md]: wow, why I could be having conflicts from my last PR that already was merged (Breeze for windows documentation)? 🙄

Comment (potiuk) [reply dev/breeze/doc/BREEZE.md]: Happens sometimes, when there som whitespace differences etc. I tend to get similar artifacts from time to time - repeated lines. Usually there some results of resolving conflicts slightly differently in two branches.

File: dev/breeze/doc/BREEZE.md
Comment (edithturn) [reply dev/breeze/doc/BREEZE.md]: I got it!  It means I didn't execute pre-commit for this, if I had done this it would catch the errors before pushing them. 

I should remember as a rule, run pre-commit always

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: Will these common functions ever get rewritten in python? or they are no longer necessary

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: yeah, I am working on this now

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [reply scripts/in_container/run_resource_check.py]: Not needed - mostly of the time (or we already use stuff instead - like "run_command"). Anything that will be needed will end up as a shared Python code.

File: dev/breeze/doc/BREEZE.md
Comment (potiuk) [reply dev/breeze/doc/BREEZE.md]: Oh yeah. Just `pre-commit install` and it will run always when you commit.

**Comment by edithturn** [comment]
@potiuk I corrected your observations, I will appreciate your feedback when you have free time :)

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]:
```suggestion
            console.print(f" * {resource} available {available[0]}{check}. [green]OK.")
```

Comment (potiuk) [reply scripts/in_container/run_resource_check.py]: NIT. It will look nicer without separate line:

image

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]: The numbers (4,2,20) are repeated here and above - they should be extracted as constants.

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]:
```suggestion
    resources= {
        'Memory': [mem_available, MAX_MEMORY]
        'Cpus' : ....
```

This would be far simpler and shorter (no need to use `setdefault` here

Comment (potiuk) [reply scripts/in_container/run_resource_check.py]: Actually it would be better to use a dataClass instead of array:

```
@dataclass
class Resource:
   current: float
   max: float
```

and then

```
resources: Dict[str, Resource] = {
     'Memory': Resource(current=mem_available, max=MAX_MEMORY),
     'Cpus': ...
}
```

Then you could refer to it via:

```
available.current, available.max
```
rather than

```
available[0], available[1]
```

**Review by potiuk — CHANGES_REQUESTED** [review]
Few simple improvements for clarity proposed.

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: Done :), definitely looks great without an end line at the end.
Screenshot from 2022-01-12 10-08-00

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: Should  I make the same with "Memory", "Cpus" and "Disk", they are used in both functions, the global variable will be good here?

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: I changed for dataclass @potiuk, I tried to give the variables an appropriate name.

**Comment by edithturn** [comment]
@potiuk this si how this script is showing the information with the job: 
Screenshot from 2022-01-12 15-00-15

However, there is an error that I do not understand yet, related to:

The BREEZE variable is not set. Defaulting to a blank string.
ERROR: for airflow "Container" 00543b681b2e "is unhealthy.
   Encountered errors while bringing up the project. " 
   
In the action: https://github.com/apache/airflow/runs/4793789902?check_suite_focus=true

**Comment by potiuk** [comment]
> The BREEZE variable is not set. Defaulting to a blank string.
> ERROR: for airflow "Container" 00543b681b2e "is unhealthy.

No no , that's fine. The tests failed randomly :)

**Comment by potiuk** [comment]
Just rebase and I am ready to merge it :)

**Comment by edithturn** [comment]
I rebased it 
image, I am not sure if this is good, because I still having BREEZE.md and the rebase --continue didn't ask me to squash. 
I need to master rebase, I am just following these instructions: https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#how-to-rebase-pr

Please Jarek, let me know if something wrong here!

**Comment by edithturn** [comment]
Let me know about this if I need to rebase it again :)

File: dev/breeze/doc/BREEZE.md
Comment (potiuk) [inline dev/breeze/doc/BREEZE.md]: I think this one should be removed completely ?

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]: I do not think we need the `resource == "..."` her e at all . This if could lbe reduced to simpy `if capacity.current < capacity.minimumAllowed:`

Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: I totally agree! Thank you!

File: scripts/in_container/run_resource_check.py
Comment (potiuk) [inline scripts/in_container/run_resource_check.py]: This will be much more accurate and easier to verify:
```suggestion
    factor = 1024*1024*1024
```

Also the get_size is in `bytes` not kilobytes ?

**Comment by potiuk** [comment]
I think there is still something wrong:

When you look at your branch (by clicking on the branch name at the top of your PR): https://github.com/edithturn/airflow/tree/verify-enough-resources-for-breeze :

| This branch is 19 commits ahead, 30 commits behind apache:main.

What you really want is that your branch is "19 commits ahead"

**Comment by edithturn** [comment]
@potiuk my main, right now have the changes of my branch "check-free-space-python" to check if build-image was working :) .
Should I reset now, right?  To work normally and fix this branch. 

> git push -f edithturn main -> to reset it back to the "real" main.

**Comment by potiuk** [comment]
`git fetch apache && git reset --hard apache/main` should reset your local main to the "apache" main.

**Comment by edithturn** [comment]
@potiuk I fixed. Now  "This branch is 3 commits ahead of apache:main."
Thank you for your patience Jarek.

File: dev/breeze/doc/BREEZE.md
Comment (edithturn) [reply dev/breeze/doc/BREEZE.md]: Thank you, I did this time.

File: scripts/in_container/run_resource_check.py
Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: I was wondering why don't write directly: 1073741824 instead of 1024*1024*1024

Comment (potiuk) [reply scripts/in_container/run_resource_check.py]: It's for readability. It's much clearer to see that it's 1024 ** 3 (you could also write it this way but multiplication is likely faster than power of). 

Simple question: Do you know by heart that 1073741824 is (as it should be) 1024 ^ 3 when you look at it? Can you verify it when code reviewing without using calculator? How about `1024*1024*1024` ? 

For me - I can immediately see that the second one is 1024 ** 3 as expected.

Comment (edithturn) [reply scripts/in_container/run_resource_check.py]: Yes, this is definitely more understandable.

@potiuk some points I checked here:

**Point 01**: gsutil returns the size of resources in bytes of memory(available) and disk(free), and the function get_size takes bytes and returns Gb.
` 1 Gigabytes = 1024*1024*1024 = 1073741824 bytes`. Here is where I am taking just the integer value:
`value_gb = bytes // factor` if this is 4.605457306 I will take 4 because rounding to 5 is not convenient, it would be to complete the missing memory, which is wrong.

**Point 02**: One of the things I realized in the previous code **run_resource_check.sh** is that in "human_readable_memory" we were storing the total memory in our system instead of Total Memory Available. Now I am using:  `mem_available = get_size(svmem.available)`


Screenshot from 2022-01-17 21-47-54


**This is to my own understanding** :)
MemFree != MemAvailable. We are using MemAvailable which is the case.
- MemFree: The amount of physical RAM, in kilobytes, left unused by the system.
- MemAvailable: An estimate of how much memory is available for starting new applications, without swapping

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
The PROD image fail is because of the problem with GPG key we fixed in `main` already. Ready to merge :)

**Comment by potiuk** [comment]
Fantastifc one @edithturn ! The "available" choice is good idea! I just tested it and ii workes really nicely so I am merging it now. 

70 less lines of BASH ! Yay :tada: :tada: :tada: :tada: !

**Comment by edithturn** [comment]
Woho! Thank youuuuuu!!! 🎊

**Comment by potiuk** [comment]
Hey @edithturn - looks like the "self-hosted" runners (with a different filesystem - in memory) had some problems so you will need to re-create the PR and we should test it there as well :(

**Comment by ashb** [comment]
Sorry to revert rather than fix -- but everything was broken on main as a result.

**Comment by potiuk** [comment]
Yeah. Good call! It was for committers only but that's "enough"

**Comment by edithturn** [comment]
Thank you so much @ashb and @potiuk, I will check the  "filesystem - in memory" in our runners, and test based on that.

**Comment by potiuk** [comment]
It was not at all obvious, really  I think the reason is this:

https://psutil.readthedocs.io/en/latest/#disks
| disk_partitions(all=False)
| If all parameter is False it tries to distinguish and return physical devices only (e.g. hard disks, cd-rom drives, USB keys) and ignore all others (e.g. pseudo, memory, duplicate, inaccessible filesystems)

Which precisely what happened. 

I think you should list all partitions,  and check the size of the partition that has "/" as mountpoint.

**Comment by edithturn** [comment]
I got it @potiuk I assumed that the first one in the array will always be the mounted point: Like in my local machine. I didn't identify it when it ran as a workflow, maybe because there were other issues. (I will check the jobs right now)

```
>>> print(partitions)
[sdiskpart(device='/dev/sda3', mountpoint='/', fstype='ext4', opts='rw,r
```
Screenshot from 2022-01-19 08-23-30

I should have taken this directly :` partition_usage = psutil.disk_usage('/')`

I will test it again!

**Comment by potiuk** [comment]
I will also push it as well when you finish to test "committer" workflow.

**Comment by potiuk** [comment]
BTW. The root cause is that In our self-hosted runners, Docker is mounted on "tmpfs" (big ram-disk) to make things faster. So the 'smart' disk_partitions() did not return any partition :shrug:

**Comment by edithturn** [comment]
> Hey @edithturn - looks like the "self-hosted" runners (with a different filesystem - in memory) had some problems so you will need to re-create the PR and we should test it there as well :(
@jarek where did you get these logs? could you point o me, thanks in advance.
I didn't find it in the workflow after it was merged.

**Comment by potiuk** [comment]
You can see link in the PR of Ash https://github.com/apache/airflow/pull/20948

**Pull request merged**
